libhttpserver/server.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2018 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libhttpserver
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/hex"
+	"errors"
+	"io"
+	"net/http"
+	"path"
+	"strings"
+
+	"github.com/hashicorp/golang-lru"
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/kbfs/libfs"
+	"github.com/keybase/kbfs/libkbfs"
+	"github.com/keybase/kbfs/libmime"
+	"github.com/keybase/kbfs/tlf"
+)
+
+const tokenCacheSize = 64
+const fsCacheSize = 64
+
+// Server is a local HTTP server for serving KBFS content to the front end.
+type Server struct {
+	config libkbfs.Config
+	server *libkb.HTTPSrv
+
+	tokens *lru.Cache
+	fs     *lru.Cache
+}
+
+var _ libkbfs.LocalHTTPServer = (*Server)(nil)
+
+const tokenByteSize = 16
+
+// NewToken implements the libkbfs.LocalHTTPServer interface..
+func (s *Server) NewToken() (token string, err error) {
+	buf := make([]byte, tokenByteSize)
+	if _, err = rand.Read(buf); err != nil {
+		return "", err
+	}
+	token = hex.EncodeToString(buf)
+	s.tokens.Add(token, nil)
+	return token, nil
+}
+
+func (s *Server) handleInvalidToken(w http.ResponseWriter) {
+	w.WriteHeader(http.StatusForbidden)
+	io.WriteString(w, `
+    <html>
+        <head>
+            <title>KBFS HTTP Token Invalid</title>
+        </head>
+        <body>
+            token invalid
+        </body>
+    </html>
+    `)
+}
+
+func (s *Server) handleBadRequest(w http.ResponseWriter) {
+	w.WriteHeader(http.StatusBadRequest)
+}
+
+func (s *Server) getHTTPFileSystem(ctx context.Context, requestPath string) (
+	toStrip string, fs http.FileSystem, err error) {
+	fields := strings.Split(requestPath, "/")
+	if len(fields) < 3 {
+		return "", nil, errors.New("bad path")
+	}
+
+	tlfType, err := tlf.ParseTlfType(fields[0])
+	if err != nil {
+		return "", nil, err
+	}
+
+	tlfHandle, err := libkbfs.GetHandleFromFolderNameAndType(ctx,
+		s.config.KBPKI(), s.config.MDOps(), fields[1], tlfType)
+	if err != nil {
+		return "", nil, err
+	}
+
+	toStrip = path.Join(fields[0], fields[1])
+
+	fav := tlfHandle.ToFavorite()
+	if fsCached, ok := s.fs.Get(fav); ok {
+		if tlfFS, ok := fsCached.(*libfs.FS); ok {
+			return toStrip, tlfFS.ToHTTPFileSystem(ctx), nil
+		}
+	}
+
+	tlfFS, err := libfs.NewFS(ctx,
+		s.config, tlfHandle, "", "", keybase1.MDPriorityNormal)
+	if err != nil {
+		return "", nil, err
+	}
+	s.fs.Add(fav, tlfFS)
```

Comment:
We might need to add hooks to delete or rename FSes from the favorites list when the handle changes.  For example, if an old TLF `a,b@rooter` becomes a conflict folder, or a subteam gets renamed (and a new subteam gets made with the old name).  You can subscribe with KBFSOps to be an observer of handle change callbacks.

Reply:
(I guess that change could be in a separate PR if you want, but we should probably do it sooner rather than later.)

Reply:
Good point; Thanks! I'll do it in a follow-up PR soon.